### PR TITLE
chore: update redis to new design (MAPCO-3769)

### PR DESCRIPTION
**Pull request by razbroc**
| Question                | Answer                                                                          |
| ---------------- | -------------------------------------------------------------------------- |
| Bug fix         | ✔                                                                        |
| New feature     | ✔                                                                        |
| Breaking change | ✔                                                                        |
| Deprecations    | ✖                                                                        |
| Documentation   | ✖                                                                        |
| Tests added     | ✖                                                                        |
| Chore            | ✔                                                                       |

Related issues: (MAPCO-3769)

### CODE REVIEW

**Review by asafMasa — CHANGES_REQUESTED** [review]
please check my remarks & run E2E for new & update ingest & cleanup for both

File: src/layers/models/layersManager.ts [hunk 1]
```diff
@@ -159,45 +161,45 @@ class LayersManager {
...
       } else {
-        newLayer = this.getLayerValues(sourceLayerName);
-        const sourceLayerNameIndex: number = jsonDocument.layers.findIndex((layer) => layer.name === sourceLayerName);
+        newLayer = this.getLayerValues(mainLayerName);
+        const sourceLayerNameIndex: number = jsonDocument.layers.findIndex((layer) => layer.name === mainLayerName);
         jsonDocument.layers[sourceLayerNameIndex] = newLayer;
       }
       const newSourceCache: IMapProxyCache = this.getCacheValues(layerRequest.cacheType, layerRequest.tilesPath, tileMimeFormat);
-      jsonDocument.caches[sourceLayerName] = newSourceCache;
+      jsonDocument.caches[mainLayerName] = newSourceCache;
```
[the remark this answers is not in the document]
Comment (asafMasa) [reply src/layers/models/layersManager.ts:202]: fix

File: src/layers/routes/layersRouterFactory.ts [hunk 2]
```diff
@@ -10,8 +10,8 @@ const layersRouterFactory: FactoryFunction<Router> = (dependencyContainer) => {
   router.post('/layer', controller.addLayer);
   router.put('/layer/:name', controller.updateLayer);
   router.delete('/layer', controller.removeLayer);
-  router.post('/mosaic/:name', controller.addLayerToMosaic);
-  router.put('/mosaic/:name', controller.updateMosaic);
+  // router.post('/mosaic/:name', controller.addLayerToMosaic);
```
[the remark this answers is not in the document]
Comment (asafMasa) [reply src/layers/routes/layersRouterFactory.ts:13]: fix

**Review by asafMasa — CHANGES_REQUESTED** [review]

File: tests/unit/layers/models/layersManager.spec.ts [hunk 3]
```diff
@@ -1,30 +1,24 @@
...
-  IMapProxyConfig,
-  IRedisConfig,
-  IUpdateMosaicRequest,
-} from '../../../../src/common/interfaces';
+import { ILayerPostRequest, IMapProxyCache, IMapProxyConfig, IRedisConfig } from '../../../../src/common/interfaces';
 import { LayersManager } from '../../../../src/layers/models/layersManager';
 import { mockLayerNameAlreadyExists } from '../../mock/mockLayerNameAlreadyExists';
 import { mockLayerNameIsNotExists } from '../../mock/mockLayerNameIsNotExists';
-import * as utils from '../../../../src/common/utils';
+// import * as utils from '../../../../src/common/utils';
```
Comment (asafMasa) [inline tests/unit/layers/models/layersManager.spec.ts:14]: ?

File: tests/unit/layers/models/layersManager.spec.ts [hunk 4]
```diff
@@ -1,30 +1,24 @@
...
 import { MockConfigProvider, getJsonMock, updateJsonMock, init as initConfigProvider } from '../../mock/mockConfigProvider';
 import { SERVICES } from '../../../../src/common/constants';
 import { registerTestValues } from '../../../integration/testContainerConfig';
 import { init as initConfig, clear as clearConfig } from '../../../configurations/config';
 
 let layersManager: LayersManager;
-let sortArrayByZIndexStub: jest.SpyInstance;
+// let sortArrayByZIndexStub: jest.SpyInstance;
```
Comment (asafMasa) [inline tests/unit/layers/models/layersManager.spec.ts:21]: ?

File: tests/unit/layers/models/layersManager.spec.ts [hunk 5]
```diff
@@ -37,7 +31,7 @@ describe('layersManager', () => {
     const redisConfig = container.resolve<IRedisConfig>(SERVICES.REDISCONFIG);
     const mapproxyConfig = container.resolve<IMapProxyConfig>(SERVICES.MAPPROXY);
     layersManager = new LayersManager(logger, mapproxyConfig, redisConfig, MockConfigProvider);
-    sortArrayByZIndexStub = jest.spyOn(utils, 'sortArrayByZIndex').mockReturnValueOnce(['mockLayer1', 'mockLayer2', 'mockLayer3']);
+    // sortArrayByZIndexStub = jest.spyOn(utils, 'sortArrayByZIndex').mockReturnValueOnce(['mockLayer1', 'mockLayer2', 'mockLayer3']);
```
Comment (asafMasa) [inline tests/unit/layers/models/layersManager.spec.ts:34]: ?

**Review by asafMasa — CHANGES_REQUESTED** [review]

File: openapi3.yaml [hunk 6]
```diff
@@ -193,107 +204,49 @@ paths:
...
+          description: OK
           content:
             application/json:
               schema:
-                $ref: '#/components/schemas/mosaic'
-        '204':
-          description: Updated
-        '400':
-          description: Bad Request
+                $ref: '#/components/schemas/getLayerResponse'
```
Comment (asafMasa) [inline openapi3.yaml:230]: please add cacheName to repsonse

File: src/common/interfaces.ts [hunk 7]
```diff
@@ -119,12 +119,16 @@ export interface IRedisSource extends ICacheSource {
   default_ttl: number;
 }
 
+export interface ICacheName {
```
Comment (asafMasa) [inline src/common/interfaces.ts:122]: fix Prettier

File: src/layers/models/layersManager.ts [hunk 8]
```diff
@@ -43,79 +42,64 @@ class LayersManager {
...
-    const editJson = (jsonDocument: IMapProxyJsonDocument): IMapProxyJsonDocument => {
-      this.addNewCache(jsonDocument, layerRequest);
-
-      return jsonDocument;
-    };
-
-    await this.configProvider.updateJson(editJson);
-  }
+  public async getCacheName(layerName: string, cacheType: string): Promise<ICacheName> {
+    const errorMsg = `Layer ${layerName} or fitting cache do not exist`
```
Comment (asafMasa) [inline src/layers/models/layersManager.ts:46]: please move this errorMsg definition to were it's beeing used

File: src/layers/models/layersManager.ts [hunk 9]
```diff
@@ -43,79 +42,64 @@ class LayersManager {
...
+    if (!this.isCacheTypeValid(cacheType)) {
+      const invalidCacheTypeMsg = `Invalid cache type: ${cacheType} has been provided , available values: "geopackage", "s3", "file", "redis"`
+      throw new Error(invalidCacheTypeMsg);
```
Comment (asafMasa) [inline src/layers/models/layersManager.ts:49]: Please add to OpenAPI scheme and return 400

File: src/layers/models/layersManager.ts [hunk 10]
```diff
@@ -43,79 +42,64 @@ class LayersManager {
...
-      // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
-      const mosaicCache: IMapProxyCache = jsonDocument.caches[mosaicName];
-      mosaicCache.sources.push(layerToMosaicRequest.layerName);
-      return jsonDocument;
-    };
-
-    await this.configProvider.updateJson(editJson);
-    this.logger.info(`Successfully added layer: '${layerToMosaicRequest.layerName}' to mosaic: '${mosaicName}'`);
+    }
+    throw new NotFoundError(errorMsg)
```
Comment (asafMasa) [inline src/layers/models/layersManager.ts:68]: log warn with the not found error and request params

File: src/layers/models/layersManager.ts (hunk 9 above)
Comment (asafMasa) [reply src/layers/models/layersManager.ts:49]: log warn with the 400 error and request params

File: src/layers/models/layersManager.ts [hunk 11]
```diff
@@ -43,79 +42,64 @@ class LayersManager {
...
 
-  public async updateMosaic(mosaicName: string, updateMosaicRequest: IUpdateMosaicRequest): Promise<void> {
-    this.logger.info(`Update mosaic: ${mosaicName} request`);
-
+  public async addLayer(layerRequest: ILayerPostRequest): Promise<void> {
     const editJson = (jsonDocument: IMapProxyJsonDocument): IMapProxyJsonDocument => {
-      if (!isLayerNameExists(jsonDocument, mosaicName)) {
-        throw new NotFoundError(`Mosaic name '${mosaicName}' is not exists`);
+      if (isRedisCacheLayer(layerRequest.name)) {
+        const errorMsg = `layer names that ends with '-redis' are not supported`;
```
Comment (asafMasa) [inline src/layers/models/layersManager.ts:74]: log warn with the NotImplementedError and request params

File: src/layers/models/layersManager.ts [hunk 12]
```diff
@@ -43,79 +42,64 @@ class LayersManager {
...
+        throw new NotImplementedError(errorMsg);
       }
-      updateMosaicRequest.layers.forEach((layer) => {
-        if (!isLayerNameExists(jsonDocument, layer.layerName)) {
-          throw new NotFoundError(`Layer name '${layer.layerName}' is not exists`);
-        }
-      });
 
-      const sortedLayers: string[] = sortArrayByZIndex(updateMosaicRequest.layers);
+      this.addNewCache(jsonDocument, layerRequest);
```
Comment (asafMasa) [inline src/layers/models/layersManager.ts:78]: check if "addLayer" method is logged

File: src/layers/models/layersManager.ts [hunk 13]
```diff
@@ -43,79 +42,64 @@ class LayersManager {
...
 
-  public getAllLinkedCaches(baseCacheNames: string[], mapproxyConfiguration: IMapProxyJsonDocument): string[] {
+  public getAllLayerLinkedCaches(baseCacheNames: string[], mapproxyConfiguration: IMapProxyJsonDocument): string[] {
     const linkedCaches: string[] = [];
     const baseCacheNamesDuplicate: string[] = [...baseCacheNames];
 
     baseCacheNamesDuplicate.forEach((currentCache) => {
-      const sourceName = currentCache.endsWith('-source') ? currentCache : `${currentCache}-source`;
-      const redisSourceName = currentCache.endsWith('-source') ? currentCache.replace('-source', '') : currentCache;
+      const mainCacheName = isRedisCacheLayer(currentCache) ? currentCache.replace('-redis', '') : currentCache;
```
Comment (asafMasa) [inline src/layers/models/layersManager.ts:91]: all '-redis' methods should be in utils same as in here -
https://github.com/MapColonies/mapproxy-api/pull/126/files#diff-a404995816c64ab594401ce132110c5ea7926a0e27d6bc19cd0a59a866ece081R45

File: src/layers/models/layersManager.ts [hunk 14]
```diff
@@ -125,28 +109,50 @@ class LayersManager {
...
     const errorMessage = 'no valid layers to delete';
-    let failedLayers;
+    let failedCaches;
+    const failedLayers: string[] = [];
+    const deletedLayers: string[] = [];
+
+    layersName.forEach((layerName) => {
+      if (isRedisCacheLayer(layerName)) {
+        const errorMsg = `layer names that ends with '-redis' are not supported`;
+        throw new NotImplementedError(errorMsg);
```
Comment (asafMasa) [inline src/layers/models/layersManager.ts:119]: log warn message with request params

File: src/layers/models/layersManager.ts [hunk 15]
```diff
@@ -165,39 +171,43 @@ class LayersManager {
   public async updateLayer(layerName: string, layerRequest: ILayerPostRequest): Promise<void> {
     this.logger.info({ msg: `Update layer: '${layerName}' request`, layerRequest });
     const tileMimeFormat = mimeLookup(layerRequest.format) as TilesMimeFormat;
-    const isRedisCache = !layerName.endsWith('-source');
+    const isRedisCache = isRedisCacheLayer(layerName);
     let doesHaveRedisCache = false;
 
+    if (isRedisCacheLayer(layerName)) {
+      const errorMsg = `layer names that ends with '-redis' are not supported`;
```
Comment (asafMasa) [inline src/layers/models/layersManager.ts:178]: add warn log with request params

File: src/layers/models/layersManager.ts [hunk 16]
```diff
@@ -266,19 +281,19 @@ class LayersManager {
...
   private addNewSourceLayerToConfig(layerRequest: ILayerPostRequest, jsonDocument: IMapProxyJsonDocument): void {
     //creates source cache, and a source layer
-    this.logger.info({ msg: `adding ${layerRequest.name} as source layer`, layerRequest });
-    const sourceCacheTitle = `${layerRequest.name}-source`;
-    if (isLayerNameExists(jsonDocument, sourceCacheTitle)) {
-      throw new ConflictError(`Layer name '${sourceCacheTitle}' already exists`);
+    this.logger.info({ msg: `adding ${layerRequest.name} as main layer`, layerRequest });
+    const mainCacheTitle = layerRequest.name;
+    if (isLayerNameExists(jsonDocument, mainCacheTitle)) {
+      throw new ConflictError(`Layer name '${mainCacheTitle}' already exists`);
```
Comment (asafMasa) [inline src/layers/models/layersManager.ts:287]: log the error

File: src/layers/routes/layersRouterFactory.ts [hunk 17]
```diff
@@ -7,12 +7,10 @@ const layersRouterFactory: FactoryFunction<Router> = (dependencyContainer) => {
   const controller = dependencyContainer.resolve(LayersController);
 
   router.get('/layer/:name', controller.getLayer);
+  router.get('/layer/:name/:type', controller.getLayersCache);
```
Comment (asafMasa) [inline src/layers/routes/layersRouterFactory.ts:10]: change to OpenAPI remark

**Review by asafMasa — CHANGES_REQUESTED** [review]
Please mind my request changes

File: openapi3.yaml [hunk 18]
```diff
@@ -193,107 +204,49 @@ paths:
...
+        '404':
+          description: Not Found
           content:
             application/json:
               schema:
                 $ref: '#/components/schemas/error'
-        '404':
-          description: Not found
+        '400':
+          description: Not valid request
```
Comment (asafMasa) [inline openapi3.yaml:238]: add supported cahche types enum

File: src/layers/models/layersManager.ts [hunk 19]
```diff
@@ -43,79 +42,64 @@ class LayersManager {
...
-  public async addLayer(layerRequest: ILayerPostRequest): Promise<void> {
-    const editJson = (jsonDocument: IMapProxyJsonDocument): IMapProxyJsonDocument => {
-      this.addNewCache(jsonDocument, layerRequest);
-
-      return jsonDocument;
-    };
-
-    await this.configProvider.updateJson(editJson);
-  }
+  public async getCacheName(layerName: string, cacheType: string): Promise<ICacheName> {
```
Comment (asafMasa) [inline src/layers/models/layersManager.ts:45]: getCacheName -> getCacheForNameAndType

File: src/layers/models/layersManager.ts [hunk 20]
```diff
@@ -43,79 +42,64 @@ class LayersManager {
...
-  public async updateMosaic(mosaicName: string, updateMosaicRequest: IUpdateMosaicRequest): Promise<void> {
-    this.logger.info(`Update mosaic: ${mosaicName} request`);
-
+  public async addLayer(layerRequest: ILayerPostRequest): Promise<void> {
     const editJson = (jsonDocument: IMapProxyJsonDocument): IMapProxyJsonDocument => {
-      if (!isLayerNameExists(jsonDocument, mosaicName)) {
-        throw new NotFoundError(`Mosaic name '${mosaicName}' is not exists`);
+      if (isRedisCacheLayer(layerRequest.name)) {
+        const errorMsg = `layer names that ends with '-redis' are not supported`;
+        throw new NotImplementedError(errorMsg);
```
Comment (asafMasa) [inline src/layers/models/layersManager.ts:75]: NotSupportedError

File: src/layers/models/layersManager.ts (hunk 14 above)
Comment (asafMasa) [reply src/layers/models/layersManager.ts:119]: NotSupportedError

**Review by asafMasa — CHANGES_REQUESTED** [review]

File: openapi3.yaml [hunk 21]
```diff
@@ -193,107 +204,49 @@ paths:
           application/json:
             schema:
               $ref: '#/components/schemas/layer'
-  /mosaic/{name}:
-    post:
-      operationId: addLayerToMosaic
+  /layer/{name}/{type}:
```
[the remark this answers is not in the document]
Comment (ronenkapelian) [reply openapi3.yaml:207]: done

File: openapi3.yaml (hunk 6 above)
Comment (ronenkapelian) [reply openapi3.yaml:230]: added new response type for cacheObject

File: openapi3.yaml (hunk 18 above)
Comment (ronenkapelian) [reply openapi3.yaml:238]: added

File: src/common/interfaces.ts (hunk 7 above)
Comment (ronenkapelian) [reply src/common/interfaces.ts:122]: fixed

File: src/layers/routes/layersRouterFactory.ts (hunk 17 above)
Comment (ronenkapelian) [reply src/layers/routes/layersRouterFactory.ts:10]: changed

File: src/layers/models/layersManager.ts (hunk 19 above)
Comment (ronenkapelian) [reply src/layers/models/layersManager.ts:45]: changed

File: src/layers/models/layersManager.ts (hunk 8 above)
Comment (ronenkapelian) [reply src/layers/models/layersManager.ts:46]: moved

File: src/layers/models/layersManager.ts (hunk 9 above)
Comment (ronenkapelian) [reply src/layers/models/layersManager.ts:49]: done

File: src/layers/models/layersManager.ts (hunk 10 above)
Comment (ronenkapelian) [reply src/layers/models/layersManager.ts:68]: done

File: src/layers/models/layersManager.ts (hunk 13 above)
Comment (ronenkapelian) [reply src/layers/models/layersManager.ts:91]: done

File: src/layers/models/layersManager.ts (hunk 14 above)
Comment (ronenkapelian) [reply src/layers/models/layersManager.ts:119]: done

File: src/layers/models/layersManager.ts (hunk 15 above)
Comment (ronenkapelian) [reply src/layers/models/layersManager.ts:178]: done

File: src/layers/models/layersManager.ts (hunk 16 above)
Comment (ronenkapelian) [reply src/layers/models/layersManager.ts:287]: done

File: src/layers/models/layersManager.ts [hunk 22]
```diff
@@ -85,12 +70,46 @@ class LayersManager {
     await this.configProvider.updateJson(editJson);
   }
 
+  public async getCacheByNameAndType(layerName: string, cacheType: string): Promise<ICacheObject> {
```
Comment (asafMasa) [inline src/layers/models/layersManager.ts:73]: add @withSpanAsyncV4

**Review by asafMasa — CHANGES_REQUESTED** [review]

**Comment by asafMasa** [comment]
moved to new PR

**Pull request closed without merging**